### ISSUE TITLE
Warning when delegating to mock in Ruby 2.4

### ISSUE THREAD

**Issue by tjvc** [freerange/mocha#321]
Ruby 2.4 warns when delegating to a private method with `Forwardable`: https://bugs.ruby-lang.org/projects/ruby-trunk/repository/revisions/56210.

When delegating to a Mocha mock object with `Forwardable`, a "forwarding to private method" warning is shown.

For example, given:

```ruby
require "forwardable"

class Foo
  extend Forwardable

  attr_reader :baz

  def_delegator :baz, :bar

  def initialize(baz)
    @baz = baz
  end
end
```

```ruby
require "minitest/autorun"
require "mocha/mini_test"

require_relative "foo"

class FooTest < Minitest::Test
  def test_delegates_bar_to_baz
    baz = mock
    baz.expects(:bar)

    foo = Foo.new(baz)
    foo.bar
  end
end
```

The test will output:

```
foo_test.rb:12:in `test_delegates_bar_to_baz': Foo#bar at /Users/thomcarter/code/mocha/foo.rb:8 forwarding to private method Mocha::Mock#bar
```

Ideally these warnings would not be shown, because in most cases the expectation is set on a public method.

**Comment by floehopper** [comment]
@tjvc:

Thanks for reporting this with such a clear example.

As far as I can see this is a problem with Ruby and not Mocha, because I see the same behaviour when I replace the mock object with a simple Ruby class using `method_missing`:

* `baz.rb`

```ruby
 class Baz
   def method_missing(symbol, *args)
     if symbol == :bar
       123
     else
       super
     end
   end
 
   def respond_to_missing?(name, include_private = false)
     (symbol == :bar) || super
   end 
 end
```

* `foo_test.rb`

```ruby
require "minitest/autorun"

require_relative "foo"
require_relative "baz"

class FooTest < Minitest::Test
  def test_delegates_bar_to_baz
    baz = Baz.new

    foo = Foo.new(baz)
    foo.bar
  end
end
```

```
foo_test.rb:11:in `test_delegates_bar_to_baz': Foo#bar at foo.rb:8 forwarding to private method Baz#bar
```

Does that make sense? I think the next step would be to report the issue to Ruby - are you happy to do that or would you like me to do it?

**Comment by floehopper** [comment]
FWIW I think the problem lies on [this line of `forwardable.rb`][1] - I'm not familiar with this use of `defined?`, but I assume it doesn't take methods defined by `method_missing` into account.

[1]: https://github.com/ruby/ruby/blob/b9bedc2d3316fc910522d399c270a565bab3d5fd/lib/forwardable.rb#L208

**Comment by floehopper** [comment]
Just to clarify, I would consider `Baz#bar` to be a "public" method, because:

```
baz = Baz.new
p baz.respond_to?(:bar, include_private = false) # => true
```

**Comment by tjvc** [comment]
Thanks for getting back to me, @floehopper.

I did a bit more digging: it looks like [an issue was already reported for Ruby](https://bugs.ruby-lang.org/issues/13326), with the suggested fix being to implement `respond_to_missing?` where `method_missing` is used. I noticed that you did this in your example, but that `respond_to_missing?` is not implemented in `Mocha::Mock`. Once I implemented it, the warnings were no longer shown.

I opened a pull request with the change (#323), but I can see that this has broken compatibility with Ruby 1.8.7, where `respond_to_missing?` is not available. Implementing both `respond_to?` and `respond_to_missing?` does not remove the warnings, so I'm not sure how best to proceed. I hope this is of some help, all the same!
